### PR TITLE
Gpg additions master

### PR DESCRIPTION
**Pull request by github-abcde**
### What does this PR do?
It adds a gpg state module, and cleans up the gpg execution module bigtime. It also adds loads of unit- and integration tests.
Things cleaned up in the gpg execution module:
- Made function return values consistently return a dict (where appropriate).
- Add all parameters in function docstrings and ensure proper python Sphinx domain formatting.
- Replace incorrect GPG_1_3_1 version check with version checks that distinguish between the version of python's gnupg library and the version of the installed GPG tools.
- Add `gnupghome`-parameter to all relevant functions.
- Adds passphrase usage options, for example when creating/importing/exporting secret keys with GPG 2.1 or later, a passphrase is now required.
- Improved passphrase-from-pillar support. Now no longer requires a fixed pillar key `gpg_passphrase`, but the pillar key can now be passed as argument `passphrase_pillar`.
- Add signing/verifying with detached signatures.
- Change default keyserver to use `pool.sks-keyservers.net`.
- Add symmetric encrypting/decrypting support.
- A lot of errorhandling has been added.

Changes to the gpg state module:
- Updated function docstrings to contain all parameters.
- Added lots of extra tests.
- All state functions are now truly idempotent (except when idempotency cannot be determined, for example when encrypting data, it cannot be determined that existing encrypted data is the same as newly encrypted data due to the random factors used in encryption).
- Added `trusted` state to ensure keys exist at a certain trust level.
- Added `data_encrypted`, `data_decrypted`, `data_signed` and `data_verified` state functions.

### What issues does this PR fix or reference?
None that I am aware of

### Previous Behavior
A lot of functionality is missing.

### New Behavior
Data encryption/decryption/signing/verification has been extended and made more robust.
State module has been extended with encryption/decryption/signing/verification states.

### Tests written?
Verily

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by github-abcde** [comment]
It seems the testsuite defaults to using `python-gnupg` version `0.4.4` which has some bugs that cause the `integration.states.test_gpg` test to fail. Is it possible to start using version `0.4.5`?

**Review by twangboy — DISMISSED** [review]

**Comment by twangboy** [comment]
@dmurphy18 Could you review this. It's quite a large change and I seem to recall you doing something with gpg recently.

**Comment by twangboy** [comment]
@garethgreenaway Would you mind taking a look. Looks like you wrote most of the original module.

**Review by dmurphy18 — COMMENTED** [review]
The codecov report platforms tested on does not include latest Ubuntu or Debian.
Would like to see this tested against Ubuntu 18.04, Debian 10 and Centos 8, since these have later versions of gnupg and gpg-agent functionality which has given issues in the past.

**Comment by github-abcde** [comment]
@dmurphy18 Can you give me any specific things to test? (other than the things already tested in the integration tests).

**Comment by dmurphy18** [comment]
@github-abcde Wanted to make sure that the tests have been run on the platforms listed above is all, before I started reviewing, since that are later platforms with different OS version of gnupg installed

**Comment by github-abcde** [comment]
I'll debug centos7 next week :)

**Comment by github-abcde** [comment]
@dmurphy18 The tests currently failing are not caused by my code changes, so it's all yours if you want to have a look :-)

**Comment by twangboy** [comment]
@github-abcde Hey, would you mind rebasing and fixing the conflicts here?

**Comment by github-abcde** [comment]
@twangboy Done

**Comment by github-abcde** [comment]
@sagetherage Aww, why was this postponed?

**Comment by sagetherage** [comment]
it still needs to be reviewed and hasn't passed all checks prior to code freeze

**Comment by github-abcde** [comment]
@sagetherage The failing checks all fail on parts outside the changes I have made. They have for some time (at least since Mar 17). Rebasing (or rather, pressing the button [Update branch] is trivial. I'm not sure if I'm the only one able to do that, but I'd rather press the button once (and start another full cycle of tests) shortly before it gets merged, than having to press the button every time the master branch gets updated (and start a cycle of tests that nobody will use).

**Review by twangboy — CHANGES_REQUESTED** [review]

**Review by garethgreenaway — DISMISSED** [review]

**Comment by garethgreenaway** [comment]
@github-abcde Looks good, once you fix the requests from Shane and those lint failures.

**Comment by github-abcde** [comment]
@sagetherage Unfortunately, I can't fix the lint failures, as they are caused by black. If I fix them, the test that checks for changes black would cause would fail.
For example: https://github.com/saltstack/salt/pull/56283/commits/29ad6568a924a85be7cf5a09557db87cfbd2ee75#diff-b8b7c8d8ad2997fb6d709835e153d8e5R908

**Comment by sagetherage** [comment]
@github-abcde ok we'll look into the lint failures

**Comment by garethgreenaway** [comment]
@github-abcde For the lint issues just remove the commas at the end of the lines where the lint failures are being triggered.  That will resolve the issues.  I was about to fix it myself but it looks like your repo doesn't allow us to edit your pull requests.

**Comment by github-abcde** [comment]
@twangboy Can you resolve the points you opened if you deem them sufficiently handled?

File: doc/conf.py
Comment (waynew) [inline doc/conf.py:154]: Looks like this is causing the docs failure

Comment (bryceml) [reply doc/conf.py:154]: this is fixed on master I think.  This probably needs to be rebased and will then be fine.

Comment (github-abcde) [reply doc/conf.py:154]: Doesn't look like it. On the other hand, it looks like I added `ansible` and `concurrent` to fix the doc build on my end, but apparently broke it in master :)

Comment (bryceml) [reply doc/conf.py:154]: yup, that sounds correct after looking at it closer.

**Comment by github-abcde** [comment]
@dwoz Failing docs test has been resolved.

**Comment by sagetherage** [comment]
@github-abcde Windows tests are failing, so we are looking into that, ponce done we will re-run these

**Comment by github-abcde** [comment]
@sagetherage Thanks! I tried to reproduce it on my windows10 installation, but the tests do not fail there.
I briefly tried to figure out what's going on, and I suspect it tries to install a specific version of 7zip that is no longer the `latest` version. Whether that's actually what's failing the test I don't know though :-). Regardless, I could not find a connection to my changes in the gpg module and the failing windows test (and since a few days also the failing test on centos7).

**Comment by sagetherage** [comment]
@github-abcde yes, it is our tests failing, I looked this yesterday and meant to put the comment in, apologies I didn't, but yes, I will get back to this once those tests are fixed on our end :)

**Comment by sagetherage** [comment]
looks like this needs some attention and I will have to remove it from the Magnesium release cycle

**Comment by twangboy** [comment]
This needs another rebase...

File: salt/states/gpg.py
Comment (twangboy) [inline salt/states/gpg.py:441]: The comment here looks strange

**Review by twangboy — CHANGES_REQUESTED** [review]

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
@github-abcde Are you still interested in this PR ?, if unresponsive the PR shall be closed after several days.
Test cases need addressing as well as review comments, and rebasing given requirements have changed since this PR was written
Also don't notice a changelog entry

**Comment by github-abcde** [comment]
@dmurphy18 I am!
However I am disappointed in the way this MR has been handled from the side of the saltstack maintainers. In all the time this MR is open, it has mostly been waiting on input from your end, and having to rebase things _yet again_ is not the most pleasant of activities.
I will see when I can find time to go through this again.

**Comment by dmurphy18** [comment]
@github-abcde Pointing out that there were requested changes
```
[twangboy](https://github.com/twangboy) requested changes on behalf of https://github.com/orgs/saltstack/teams/team-core [on Feb 11, 2021](https://github.com/saltstack/salt/pull/56283#pullrequestreview-588624135)
```
And there are conflicts which must be resolved before continuing, perhaps you failed to notice those.

Apologize for any delays on Salt's behalf in reminding you of those issues.

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

**Pull request closed without merging**
